petastorm/pytorch.py

```diff
@@ -13,11 +13,74 @@
 # limitations under the License.
 
 # Must import pyarrow before torch. See: https://github.com/uber/petastorm/blob/master/docs/troubleshoot.rst
-import pyarrow  # noqa: F401 pylint: disable=W0611
+import collections
+import decimal
+import re
 
+import numpy as np
+from six import PY2
 from torch.utils.data.dataloader import default_collate
 
 
+def _sanitize_pytorch_types(row_as_dict):
+    """Promotes values types in a dictionary to the types supported by pytorch. Raises an error if type is clear error
+    if the type can not be promoted.
+
+    The parameter is modified in-place.
+
+    int8, uint16 are promoted to int32; uint32 -> int64;
+    numpy string_, unicode_, object arrays are not supported.
+
+    :param row_as_dict: a dictinoary of key-value pairs. The values types are promoted to pytorch compatible.
+    :return: None
+    """
+    for name, value in row_as_dict.items():
+        # PyTorch supported types are: double, float, float16, int64, int32, and uint8
+        if isinstance(value, np.ndarray):
+            if value.dtype == np.int8:
+                row_as_dict[name] = value.astype(np.int16)
```

Comment:
`int8` (and `uint8`) have native pytorch types, so they dont need upcasting

Reply:
This is what I see (`int8` appears to be unsupported, while `uint8` is supported):
```
In [1]: import numpy as np; import torch

In [2]: torch.as_tensor(np.asarray([1,2]).astype(np.int8))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-02b070359a1b> in <module>()
----> 1 torch.as_tensor(np.asarray([1,2]).astype(np.int8))

TypeError: can't convert np.ndarray of type numpy.int8. The only supported types are: double, float, float16, int64, int32, and uint8.

In [3]: torch.as_tensor(np.asarray([1,2]).astype(np.uint8))
Out[3]: tensor([1, 2], dtype=torch.uint8)
```

Reply:
You're right. Seems to be a mistake in how pytorch handles the type, considering you can create tensors of type `int8`, in any case the casting is safe so I guess it doesnt really matter.

```
In [23]: torch.CharTensor(np.zeros(1, dtype=np.int16))
Out[23]: tensor([0], dtype=torch.int8)
```